File: leetcode/top150/25.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
  public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* dummy = new ListNode(0, head);
        ListNode* p0 = dummy;
        int len = 0;
        while (head) {
            head = head->next;
            ++len;
        }
        head = dummy->next;
        for (int i = 0; i < len / k; ++i) {
            ListNode *pre, *nxt;
            for (int j = 0; j < k; ++j) {
                nxt = head->next;
                head->next = pre;
                pre = head;
                head = nxt;
            }
            nxt = p0->next;
            nxt->next = head;
            p0->next = pre;
            p0 = nxt;
        }
        return dummy->next;
    }
};

File: leetcode/hot100_python/25.py
from typing import *

class ListNode:
    def __init__(self, x=0, next=None):
        self.val = x
        self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        length = 0
        curr = head
        while curr:
            length += 1
            curr = curr.next
        dummy = ListNode(0, head)
        p0, curr = dummy, head
        while length >= k:
            length -= k
            pre, nxt = None, None
            for _ in range(k):
                nxt = curr.next
                curr.next = pre
                pre = curr
                curr = nxt
            nxt = p0.next
            nxt.next = curr
            p0.next = pre
            p0 = nxt
        return dummy.next
        